### PR TITLE
Updated block reward fomula

### PR DESCRIPTION
**Pull request by speqtr**
This is a replacement for PR #99.

### CODE REVIEW

**Comment by speqtr** [comment]
In addition to the current changes (from PR #99) it's also required to prepare code base for a soft fork in order to give everyone sometime to update to the latest version and start using new block reward formula all at the same time.

**Comment by speqtr** [comment]
@exploshot @nnian and @Xecute0101 are invited to this discussion 🙂

**Comment by nnian** [comment]
How about activating this new calculation at a height of 400,000 (as an example in the Upgrade Heights in `/src/config/CryptoNoteConfig.h`)?

**Comment by speqtr** [comment]
> How about activating this new calculation at a height of 400,000 (as an example in the Upgrade Heights in /src/config/CryptoNoteConfig.h)?

Sounds good to me! 👍

@exploshot @Xecute0101 any comments? 🙂

**Comment by nnian** [comment]
@speqtr 
Looks like ExploShot and Xecute are too busy. Then let's finish this PR.

**Comment by nnian** [comment]
I would recommend:
add this to `/src/config/CryptoNoteConfig.h`
in line 140 `const uint32_t UPGRADE_HEIGHT_REWARD_SCHEME                   = 400000;`

`lib/CryptoNoteCore/Currency.cpp`

Change `lib/CryptoNoteCore/Currency.cpp` in line 181 

```
if (height < CryptoNote::parameters::UPGRADE_HEIGHT_REWARD_SCHEME) {
    reward = penalizedBaseReward + penalizedFee;
} else {
    reward = (penalizedBaseReward + penalizedFee) * consistency;
}
```

But most annoying will be probably to ue parameter `height` in getBlockReward() hm,

**Comment by speqtr** [comment]
> But most annoying will be probably to use parameter height in getBlockReward() hm,
Yes, that is the main problem, since we don't have access to `height` in `getBlockReward` method.

**Comment by Xecute0101** [comment]
If there is not any conflicts, please merge this PR. @speqtr, any comments?

**Comment by speqtr** [comment]
@Xecute0101 I'm still looking for an option to add access to `height` value in `getBlockReward` method in order to make it possible to activate new block reward formula from particular height for all miners. (this was the reason why @nnian reverted previously merge PR with this exact same changes).

**Comment by speqtr** [comment]
My plan is to push slightly updated API with `height` support in `getBlockReward` method tomorrow.

**Comment by nnian** [comment]
> 
> 
> If there is not any conflicts, please merge this PR. @speqtr, any comments?

The delay is due to me. I think we shouldn't change the calculation of the rewards without a hard fork. Alternative would be a soft fork, so changing the calculation formula at a certain height. The users need time to update. I think there would problems if nodes calculate different rewards. But maybe i'm wrong?

**Comment by Xecute0101** [comment]
It would be safer to have a hardfork. But if we are doing a soft fork, then what we could do is.

@soft fork height
1. make an api from block explorer for the official version
2. have the daemon receive the info from the api and make it as minimum version.

We can do a swift change with most of our nodes and pools. Should not have a problem. Heavy miners will also update their pool nodes soon enough.

File: lib/CryptoNoteCore/Currency.cpp [hunk 1]
```diff
@@ -166,8 +167,18 @@ bool Currency::getBlockReward(
         penalizedFee = getPenalizedAmount(fee, medianSize, currentBlockSize);
     }
 
+    uint64_t consistency = 1;
```
Comment (aivve) [inline lib/CryptoNoteCore/Currency.cpp]: Change this to double, because with integer you will get either 0, 1 or 2.

Comment (speqtr) [reply lib/CryptoNoteCore/Currency.cpp]: Thanks for noticing this tiny bug! 🙂

Comment (speqtr) [reply lib/CryptoNoteCore/Currency.cpp]: Fixed in the latest commit in this PR.

**Comment by Xecute0101** [comment]
@speqtr, consistency requires at least two decimal points. Could you please change that?

File: lib/CryptoNoteCore/Currency.cpp [hunk 2]
```diff
@@ -152,29 +152,27 @@ bool Currency::getBlockReward(
...
-    uint64_t consistency = 1;
+    double consistency = 1.0;
     if (difficultyTarget() != 0) {
         // blockTarget is (Timestamp of New Block - Timestamp of Previous Block)
         consistency = blockTarget / difficultyTarget();
 
         // consistency range is 0..2
-        consistency = std::max<uint64_t>(consistency, 0);
-        consistency = std::min<uint64_t>(consistency, 2);
+        consistency = std::max<uint64_t>(consistency, 0.0);
```
Comment (aivve) [inline lib/CryptoNoteCore/Currency.cpp]: Change these to double too, and make casts:
```
consistency = std::max<double>(consistency, 0.0);
consistency = std::min<double>(consistency, 2.0);
```

Comment (speqtr) [reply lib/CryptoNoteCore/Currency.cpp]: Fixed in the latest commit (`7ffb667`).

File: lib/CryptoNoteCore/Currency.cpp [hunk 3]
```diff
@@ -166,17 +167,19 @@ bool Currency::getBlockReward(
...
-        consistency = std::max<uint64_t>(consistency, 0.0);
-        consistency = std::min<uint64_t>(consistency, 2.0);
+        consistency = std::max<double>(consistency, 0.0);
+        consistency = std::min<double>(consistency, 2.0);
     }
 
+    double penalizedReward = static_cast<double>(penalizedBaseReward + penalizedFee);
+
     emissionChange = penalizedBaseReward - (fee - penalizedFee);
-    reward = (penalizedBaseReward + penalizedFee) * consistency;
```
Comment (aivve) [inline lib/CryptoNoteCore/Currency.cpp]: You burn fees and don't credit them to miners?

**Comment by Xecute0101** [comment]
@aivve, 

reward = (penalizedBaseReward + penalizedFee) * consistency;

should be

reward = (penalizedBaseReward * consistency) + penalizedFee

Am I understanding you correct on this?

**Comment by aivve** [comment]
> 
> 
> @aivve,
> 
> reward = (penalizedBaseReward + penalizedFee) * consistency;
> 
> should be
> 
> reward = (penalizedBaseReward * consistency) + penalizedFee
> 
> Am I understanding you correct on this?

No everything is fine now, this is handled two lines above

**Comment by speqtr** [comment]
I think that this PR can be merged if no one has any extra comments or additional code changes requests.

cc @aivve @nnian @Xecute0101

**Pull request merged**